texar/data/data/mono_text_data.py

```diff
@@ -354,10 +411,10 @@ def process(self, raw_example: str) -> List[str]:
             if self._length_filter_mode is _LengthFilterMode.TRUNC:
                 words = words[:self._max_seq_length]
 
-        if self._bos_token != '':
-            words.insert(0, self._bos_token)
-        if self._eos_token != '':
-            words.append(self._eos_token)
+        if self._hparams.dataset["bos_token"] != '':
+            words.insert(0, self._hparams.dataset["bos_token"])
+        if self._hparams.dataset["eos_token"] != '':
+            words.append(self._hparams.dataset["eos_token"])
```

Comment:
I don't think these need to be changed, as `_bos_token` and `_eos_token` are also added as attributes in the `_construct` method.

Reply:
These were changed due to the issue I discussed with you last week, where `self.vocab["bos_token"]` and `self.vocab["eos_token"]` will be different from `hparams.dataset["bos_token"]` and `hparams.dataset["eos_token"]` in some cases. I was discussing with you the inconsistency between the two variables.

Reply:
Yes, but when `MonoTextData` is constructed using the normal `__init__` constructor, or is constructed by calling `_construct` without specifying the vocabulary, `_bos_token` is just `hparams.dataset.bos_token`, which is what we want here. What I didn't understand is why we're setting `_bos_token` to the token stored in the vocabulary, if `_construct` is called with vocabulary specified. I think it would be more clear if we unified the behaviors.

Reply:
I think unifying the two sets of variables will produce consistent behavior.

Reply:
Correct me if I'm wrong, but I thought the "inconsistent behavior" we discussed was only between hparams and the vocabulary -- `bos_token` in hparams could be `''` but is always a non-empty string in vocabulary. This is irrelevant of whether vocabularies are shared or not, the `bos_token` using in `process` should always be from the hparams. Please let me know if I understood any part of this incorrectly.

Reply:
Yes, you are right - the inconsistency is in the special tokens between `hparams` and the `vocabulary`. Previously, in `process` method of `MonoTextData`, `bos_token` was used from the vocab, I changed it be picked from `hparams` in this PR.  Having said that, I am inclined to think this inconsistency may cause some minor issues in the future.

Reply:
What I suggested here is to use `_bos_token` in `process`, and always make `_bos_token` the token from `hparams`. I agree that the inconsistency could cause troubles for the user, but I think it is inconsistent for a reason. For example, the input files might already contain some kind of BOS & EOS tokens, so there's no point in appending them in `process`, but it's required to map them in the vocabulary.

For now, I think the best thing to do is to document this behavior, and leave a note for future fixes.